File: c/secret-handshake/src/secret_handshake.c
#include <stdlib.h>
#include "secret_handshake.h"

static const char *COMMANDS[] = {
  "wink",
  "double blink",
  "close your eyes",
  "jump",
  NULL
};

static const int LEN = (sizeof(COMMANDS) / sizeof(char*)) - 1;

const char** commands(const int secret) {
  const char **result = malloc(LEN * sizeof(char*));
  const int rev = secret & (1 << LEN);
  int count = 0;

  for (int i = rev ? LEN - 1 : 0; rev ? i >= 0 : i < LEN; rev ? i-- : i++) {
      if (secret & (1 << i)) {
        result[count++] = COMMANDS[i];
      }
    }

  return result;
}

File: go/secret-handshake/secret_handshake.go
package secret

var moves = []string{
	"wink",
	"double blink",
	"close your eyes",
	"jump",
}

// isBitSet checks if a bit is set in a given number.
func isBitSet(num uint, bit uint) bool {
	return (num & (1 << bit)) > 0
}

// Handshake calculates the movements of a
// Secret Handshake according to a given code.
func Handshake(code uint) (handshake []string) {
	reverse := isBitSet(code, uint(len(moves)))

	for index, move := range moves {
		if !isBitSet(code, uint(index)) {
			continue
		}

		if reverse {
			handshake = append([]string{move}, handshake...)
		} else {
			handshake = append(handshake, move)
		}
	}

	return handshake
}
